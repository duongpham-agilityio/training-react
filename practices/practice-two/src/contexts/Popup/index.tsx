import { ReactNode, MouseEvent, useState, useMemo, useCallback } from 'react';

// Components
import { Popup } from 'components';

// Contexts
import { PopupContext } from 'contexts/Popup/context';

// Constants
import { MESSAGES } from '@constants';

// const initialState = {
//   isShown: false,
//   accepter: () => {},
// };

const PopupProvider = ({ children }: { children: ReactNode }) => {
  const [state, setState] = useState<{
    isShown: boolean;
    accepter: (_event: MouseEvent) => void;
  } | null>(null);

  const props = useMemo(() => {
    return {
      dispatch: (callback: () => void) => {
        setState({
          isShown: true,
          accepter: callback,
        });
      },
    };
  }, []);

  const onAccept = useCallback(
    (event: MouseEvent) => {
      state?.accepter(event);

      setState(null);
    },
    [state]
  );

  return (
    <PopupContext.Provider value={props}>
      {children}

      {state?.isShown && (
        <Popup
          title={MESSAGES.DELETE_TITLE}
          description={MESSAGES.DELETE_DESCRIPTION}
          onCancel={() => setState(null)}
          onAccept={onAccept}
        />
      )}
    </PopupContext.Provider>
  );
};

export default PopupProvider;
