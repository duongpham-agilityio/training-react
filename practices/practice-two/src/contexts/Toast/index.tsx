import { ReactNode, useCallback, useState } from 'react';

// Components
import { NotificationProps, Notification } from 'components';

// Context
import { ToastContext } from 'contexts/Toast/context';

const ToastProvider = ({ children }: { children: ReactNode }) => {
  const [notification, setNotification] = useState<
    NotificationProps | undefined
  >();

  const handleSetNotification = useCallback((props: NotificationProps) => {
    setNotification(props);

    setTimeout(() => {
      setNotification(undefined);
    }, 2000);
  }, []);

  return (
    <ToastContext.Provider value={{ setNotification: handleSetNotification }}>
      {children}

      {notification && <Notification {...notification} />}
    </ToastContext.Provider>
  );
};

export default ToastProvider;
