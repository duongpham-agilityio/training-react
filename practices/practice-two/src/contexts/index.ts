export { default as ToastProvider } from 'contexts/Toast';
export { default as PopupProvider } from 'contexts/Popup';
