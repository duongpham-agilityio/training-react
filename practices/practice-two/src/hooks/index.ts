export * from './useBookForm';
export * from './useSearchParam';
export * from './usePagination';
export * from './useFilter';
export * from './useDebounce';
export * from './useBooks';
export * from './useBook';
export * from './usePopupContext';
export * from './useToastContext';
