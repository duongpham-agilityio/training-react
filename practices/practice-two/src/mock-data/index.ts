export * from 'mock-data/book';
export * from 'mock-data/books';
