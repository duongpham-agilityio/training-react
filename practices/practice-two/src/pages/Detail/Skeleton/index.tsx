//Components
import { Button, Heading } from 'components/commons';

// Styles
import skeletonStyles from 'styles/commons/index.module.css';
import styles from 'pages/Detail/Skeleton/index.module.css';

const DetailSkeleton = () => {
  return (
    <section className={`${styles.detail} ${skeletonStyles.loading}`}>
      <div className={styles.detailItem}>
        <Heading
          label={''}
          size="xl"
          className={`${styles.heading} ${skeletonStyles.loadingItem}`}
        />
        <div>
          <div className={styles.action}>
            <Button className={`${styles.btn} ${skeletonStyles.loadingItem}`} />
            <Button className={`${styles.btn} ${skeletonStyles.loadingItem}`} />
            <Button
              width="w-lg"
              className={`${styles.btn} ${skeletonStyles.loadingItem}`}
            />
          </div>
        </div>
      </div>
      <div className={styles.detailItem}>
        <div className={styles.info}>
          <p
            className={`${styles.description} ${skeletonStyles.loadingItem}`}
          ></p>
          <ul className={styles.listInfo}>
            <li
              className={`${styles.infoItem} ${skeletonStyles.loadingItem}`}
            ></li>
            <li
              className={`${styles.infoItem} ${skeletonStyles.loadingItem}`}
            ></li>
            <li
              className={`${styles.infoItem} ${skeletonStyles.loadingItem}`}
            ></li>
          </ul>
        </div>
        <div className={styles.imageWrap}>
          <div
            className={`${styles.image} ${skeletonStyles.loadingItem}`}
          ></div>
        </div>
      </div>
    </section>
  );
};

export default DetailSkeleton;
