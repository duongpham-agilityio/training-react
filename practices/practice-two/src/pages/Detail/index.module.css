.detail {
  font-family: var(--primary-font);
  padding: 40px 0;
}

.detailItem {
  display: flex;
  justify-content: space-between;
  gap: 20px;
}

.detailItem:last-child {
  padding: 20px 0;
}

.heading,
.description {
  flex: 1;
}

.description {
  line-height: 36px;
}

.action {
  display: flex;
  gap: 15px;
}

.action .btn {
  width: 56px;
  height: 56px;
}

.info {
  color: var(--secondary-text-color);
  padding: 10px 0;
}

.description {
  font-size: 30px;
}

.description .actionText {
  display: inline-block;
  color: var(--secondary-color);
  padding: 0 10px;
  cursor: pointer;
}

.listInfo {
  font-size: 30px;
  padding: 45px 30px;
}

.infoItem {
  list-style: disc;
  padding: 15px 0;
}

.label {
  font-weight: 700;
}

.image {
  width: 310px;
  height: 400px;
  border-radius: 7px;
  object-fit: contain;
  background-color: var(--light-gray);
}
