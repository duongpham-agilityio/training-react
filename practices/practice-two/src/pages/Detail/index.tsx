import { memo, useCallback, useMemo, useState } from 'react';
import { useNavigate } from 'react-router-dom';

// Hooks
import { useBook } from 'hooks';

// HOCs
import { WithUseFormProps, withUseForm, withErrorBoundaries } from 'hocs';

// Components
import { Button, Heading } from 'components/commons';
import { Error } from 'components';
import DetailSkeleton from 'pages/Detail/Skeleton';

// Styles
import containerStyles from 'styles/commons/index.module.css';
import styles from 'pages/Detail/index.module.css';

// Assets
import TrashIcon from 'assets/icons/trash.svg';
import BackIcon from 'assets/icons/back.svg';
import PenCilIcon from 'assets/icons/pencil.svg';

// Helpers
import { convertDateTimeToTimeString, convertTimeToDate } from 'helpers';
import { MESSAGES } from '@constants';

const Detail = ({ dispatchAction }: WithUseFormProps) => {
  const [isShortCut, setIsShortCut] = useState(true);
  const { data, error, isLoading, deleteBook } = useBook();
  const redirect = useNavigate();
  const book = useMemo(() => {
    if (!data) return null;
    const { publishDate, createdAt, updatedAt, ...rest } = data;

    const publishDateConvert = `${convertDateTimeToTimeString(
      publishDate
    )}, ${convertTimeToDate(publishDate, '/', true)}`;

    const createdAtConvert = `${convertDateTimeToTimeString(
      createdAt
    )}, ${convertTimeToDate(createdAt, '/', true)}`;

    const updatedAtConvert = `${convertDateTimeToTimeString(
      updatedAt
    )}, ${convertTimeToDate(updatedAt, '/', true)}`;

    return {
      ...rest,
      publishDate: publishDateConvert,
      createdAt: createdAtConvert,
      updatedAt: updatedAtConvert,
    };
  }, [data]);

  const description = useMemo(() => {
    if (!data) return null;

    const { description } = data;

    return (
      <>
        {isShortCut ? (
          <span>{description.substring(0, 140)}</span>
        ) : (
          <span>{description}</span>
        )}

        {`${description}`.length > 150 && (
          <>
            {isShortCut ? (
              <span
                className={styles.actionText}
                onClick={() => setIsShortCut(false)}
              >
                show more
              </span>
            ) : (
              <span
                className={styles.actionText}
                onClick={() => setIsShortCut(true)}
              >
                show hide
              </span>
            )}
          </>
        )}
      </>
    );
  }, [isShortCut, data]);

  const navigate = useCallback(() => redirect(-1), [redirect]);

  const editHandler = useCallback(
    () =>
      data &&
      dispatchAction({
        formData: data,
        title: 'Edit book',
        type: 'update',
      }),
    [data, dispatchAction]
  );

  if (error) {
    return <Error />;
  }

  if (!data || !book)
    return (
      <main>
        <p>{MESSAGES.EMPTY_FIELD_DESCRIPTION}</p>
      </main>
    );

  return (
    <main className={containerStyles.container}>
      {isLoading ? (
        <DetailSkeleton />
      ) : (
        <section className={styles.detail}>
          <div className={styles.detailItem}>
            <Heading label={data.name} size="xl" className={styles.heading} />
            <div>
              <div className={styles.action}>
                <Button
                  leftIcon={TrashIcon}
                  variant="danger"
                  className={styles.btn}
                  onClick={deleteBook}
                  data-testid="delete-btn"
                />
                <Button
                  leftIcon={BackIcon}
                  variant="primary"
                  className={styles.btn}
                  onClick={navigate}
                  data-testid="back-btn"
                />
                <Button
                  label="Edit"
                  leftIcon={PenCilIcon}
                  variant="primary"
                  width="w-lg"
                  border="b-lg"
                  onClick={editHandler}
                  data-testid="edit-btn"
                />
              </div>
            </div>
          </div>
          <div className={styles.detailItem}>
            <div className={styles.info}>
              <p className={styles.description}>{description}</p>
              <ul className={styles.listInfo}>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Author: </span>
                  <span className={styles.text}>{book.author}</span>
                </li>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Published date: </span>
                  <span className={styles.text}>{book.publishDate}</span>
                </li>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Created at: </span>
                  <span className={styles.text}>{book.createdAt}</span>
                </li>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Updated at: </span>
                  <span className={styles.text}>{book.updatedAt}</span>
                </li>
              </ul>
            </div>
            <div className={styles.imageWrap}>
              <img
                src={data.imageURL}
                alt={data.name}
                className={styles.image}
              />
            </div>
          </div>
        </section>
      )}
    </main>
  );
};

export default memo(withErrorBoundaries(withUseForm(Detail)));
