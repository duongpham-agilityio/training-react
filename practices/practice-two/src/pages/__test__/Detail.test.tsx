// Mocks
import 'components/__test__/mocks/image.test';

import { act, fireEvent, render } from '@testing-library/react';
import * as Router from 'react-router-dom';
import { ReactNode } from 'react';

// Hooks
import * as hooks from 'hooks';

// Components
import Detail from 'pages/Detail';

// Providers
import { PopupProvider, ToastProvider } from 'contexts';

// Mocks data
import { books } from 'mock-data';

// Constants
import { MESSAGES } from '@constants';

jest.mock('hooks', () => ({
  ...jest.requireActual('hooks'),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
}));

const Wrapper = ({ children }: { children: ReactNode }) => (
  <Router.BrowserRouter>
    <ToastProvider>
      <PopupProvider>{children}</PopupProvider>
    </ToastProvider>
  </Router.BrowserRouter>
);

const setup = () =>
  render(<Detail />, {
    wrapper: Wrapper,
  });

const defaultData: ReturnType<typeof hooks.useBook> = {
  deleteBook: jest.fn(),
  data: books[0],
  error: false,
  mutate: jest.fn(),
  isValidating: false,
  isLoading: false,
};

describe('Detail page', () => {
  it('Match to snapshot', () => {
    jest.spyOn(hooks, 'useBook').mockReturnValue(defaultData);
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('Render with null data', () => {
    jest.spyOn(hooks, 'useBook').mockReturnValue({
      ...defaultData,
      data: undefined,
    });
    const { getByText } = setup();

    expect(getByText(MESSAGES.EMPTY_FIELD_DESCRIPTION)).toBeInTheDocument();
  });

  it('Render with error', () => {
    jest.spyOn(hooks, 'useBook').mockReturnValue({
      ...defaultData,
      error: true,
    });
    const { getByText } = setup();

    expect(getByText(MESSAGES.ERROR_DESCRIPTION)).toBeInTheDocument();
  });

  it('Navigate', () => {
    const navigate = jest.fn();
    jest.spyOn(hooks, 'useBook').mockReturnValue(defaultData);
    jest.spyOn(Router, 'useNavigate').mockReturnValue(navigate);
    const { getByTestId } = setup();
    const backBtn = getByTestId('back-btn');

    fireEvent.click(backBtn);

    expect(navigate).toBeCalled();
  });

  it('Edit book', () => {
    jest.spyOn(hooks, 'useBook').mockReturnValue(defaultData);
    const { getByTestId } = setup();
    const backBtn = getByTestId('edit-btn');

    act(() => {
      fireEvent.click(backBtn);
    });
  });
});
