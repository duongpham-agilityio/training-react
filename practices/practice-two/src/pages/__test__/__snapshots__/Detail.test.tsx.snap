// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail page Match to snapshot 1`] = `
<div>
  <main>
    <section>
      <div>
        <h2
          class="undefined undefined "
        >
          HTML/CSS Ebook 1
        </h2>
        <div>
          <div>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              data-testid="delete-btn"
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/trash.svg"
                width="22px"
              />
            </button>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              data-testid="back-btn"
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/back.svg"
                width="22px"
              />
            </button>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              data-testid="edit-btn"
              type="button"
            >
              <img
                alt="This is the left icon for the button"
                height="25px"
                src="assets/icons/pencil.svg"
                width="22px"
              />
              Edit
            </button>
          </div>
        </div>
      </div>
      <div>
        <div>
          <p>
            <span>
              Description of some book will displayed here
            </span>
          </p>
          <ul>
            <li>
              <span>
                Author: 
              </span>
              <span>
                Duong.Pham
              </span>
            </li>
            <li>
              <span>
                Published date: 
              </span>
              <span>
                1:39 PM, 30/03/2023
              </span>
            </li>
            <li>
              <span>
                Created at: 
              </span>
              <span>
                1:39 PM, 30/03/2023
              </span>
            </li>
            <li>
              <span>
                Updated at: 
              </span>
              <span>
                1:39 PM, 30/03/2023
              </span>
            </li>
          </ul>
        </div>
        <div>
          <img
            alt="HTML/CSS Ebook 1"
            src="http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
          />
        </div>
      </div>
    </section>
  </main>
</div>
`;
